### ISSUE TITLE
More Leagues for Fotmob's Shot Scraper

### ISSUE THREAD

**Issue by hedonistrh** [oseymour/ScraperFC#4]
Hey, I was also writing scraper for **_Fotmob_** and realized that you already pushed some changes. * 💯  When I check leagues, realized that you can also add following links into [here](https://github.com/oseymour/ScraperFC/blob/main/code/ScraperFC/FotMob.py#L39-L58) 

```python
            "Champions League":           "https://www.fotmob.com/leagues/42/matches/champions-league",
            "Euro":                       "https://www.fotmob.com/leagues/50/matches/euro",
            "Europa League":              "https://www.fotmob.com/leagues/73/matches/europa-league",
```
Those links also have shot-maps as that is what code is looking for. _(For instance **Copa America** and **Conference League** do not have that so possibly adding that does not make sense now.)_

* I'll also share my implementation when that is done. I appreciate that you made this repo though. It really inspired me. 🙏

**Comment by oseymour** [comment]
Hey Halil! It's funny you opened this. I just added those links (plus the WC) last night, I just haven't pushed the code yet. I'll close this when I do push.

I'm glad it's inspired you! And I'd love to see your implementation. The code I have right now is _really_ slow, so if you can get something that runs faster that'd be great!

**Comment by hedonistrh** [comment]
Thanks a lot for the heads-up. 👏  For the implementation, probably I'll finish that by EOW and will share with you. Also probably will profile _your code_ and _that implementation_. So we can understand how performance differs. 💪  Will update about that as well.

**Comment by hedonistrh** [comment]
My implementation to scrape shotmap from fotmob is done. I am not sure about what is the best way to share it so I am directly appending into this message. I also did how we can get top-stats etc. as well but not sharing in this message as that makes it so long and hard trace. 

In my computer _(Macbook Pro, Early 2015)_, each shot takes around 0.7 seconds _(So for one match, we have around 15-20 shots, that means we need 10-15 seconds_). That is also not-that-fast but I am not sure what are your current profiling. 🤔 Anyways, I hope that will be helpful. 🙏 

```python
from selenium import webdriver
from selenium.webdriver.support.wait import WebDriverWait
from webdriver_manager.chrome import ChromeDriverManager
from tqdm import tqdm

def open_the_url(url, driver, wait_until_id=None, headless=False):
    if driver == None:
        options = webdriver.ChromeOptions()
        options.add_argument("window-size=1400,600")
        if headless:
            options.headless = True
        options.add_argument("--auto-open-devtools-for-tabs")
        driver = webdriver.Chrome(ChromeDriverManager().install(), options=options)

    driver.get(url)
    print ("{} is opened with headless as {}".format(url, headless))
    if wait_until_id != None:
        try:
            element = WebDriverWait(driver, 10).until(lambda x: x.find_element_by_id(wait_until_id))
        except:
            raise Exception ("Timeout while we are in the wait_until")

    return driver


"""
Get team name, score, round and leagure from the header
"""
def get_meta_info_of_the_match(driver):
    import re
    header_with_score = driver.find_element_by_xpath('//*[@id="MatchFactsWrapper"]/div/div[contains(@class, "LeftColumn")]/div[contains(@class, "MFHeaderFullscreen")]/section/div[contains(@class, "MFHeaderFullscreenCSS")]/section[contains(@class, "MFHeaderFullscreenSection")]/header')
    teams_infos_in_header = header_with_score.find_elements_by_xpath('.//a')
    home_team_name = teams_infos_in_header[0].text
    home_team_url = teams_infos_in_header[0].get_attribute("href")
    home_team_id = re.findall(r'\d+', home_team_url)[0]
    home_team_logo_url = teams_infos_in_header[0].find_element_by_xpath('.//div/img').get_attribute("src")
    if home_team_id != re.findall(r'\d+', home_team_logo_url)[0]:
        raise Exception ("Id of home-team-url and home-team-logo does not match")

    away_team_name = teams_infos_in_header[1].text
    away_team_url = teams_infos_in_header[1].get_attribute("href")
    away_team_id = re.findall(r'\d+', away_team_url)[0]
    away_team_logo_url = teams_infos_in_header[1].find_element_by_xpath('.//div/img').get_attribute("src")
    if away_team_id != re.findall(r'\d+', away_team_logo_url)[0]:
        raise Exception ("Id of away-team-url and away-team-logo does not match")

    scoreboard = header_with_score.find_elements_by_xpath('.//div[contains(@class, "wrapper")]/span')
    score = scoreboard[0].text
    minute_of_the_game = scoreboard[1].text
    return {"home_team_name":home_team_name, "home_team_url":home_team_url, 
            "home_team_logo_url":home_team_logo_url, "home_team_id":home_team_id,
            "away_team_name":away_team_name, "away_team_url":away_team_url, 
            "away_team_logo_url":away_team_logo_url, "away_team_id":away_team_id,
            "score": score, "minute": minute_of_the_game}


def get_shotmap(driver, sort=True):
    """
    It will return list with dicts. Each dict has following keys and fields
    - location: {cx: ..., cy: ...} -- To show x and y location of shot in the field. And they are float.
    - type_and_result: {Shot type: ..., Situation: ..., Result: ...} 
    - player_url -- url of the player who shoot
    - player_name -- name of the player who shoot (Be careful, it will say 'De Bruyne' not 'Kevin De Bruyne')
    - minute -- minute of the shot. It is in float and 45 + 2' represented as 47. It is float.
    - minute_without_addition -- for 45 + 2', it is 45. For 47', that is 47. It is float.
    - team_id_of_player -- team id of the player who shoot. It is something like 8456 for Man City for instance. It is float.
    - team_side_of_player -- Can be 'home' or 'away'
    - xG -- xG of the goal. Float or None.
    - xGOT -- xGOT of the goal. Float or None.

    Also that list's elements are reverse in terms of when they happen. So first element is the last shot of the match.
    Because we are clicking left button while we are iterating over shot map. So it goes to 'previous shot'
    To change that (a.k.a to put first shot of the match into first element of list), we have 'sort' argument.
    That is True by definition but one can change it. 
    """
    meta_of_match = get_meta_info_of_the_match(driver)
    related_part = get_related_part_of_page(driver)
    shotmap_container = related_part.find_element_by_xpath('.//div[contains(@class, "Shotmap")]/div[contains(@class, "FullscreenShotmapContainer-commonShotMapContainer")]')
    
    # so we have two parts here, one of them is svg at the left and info about shots at the right
    info_about_field = shotmap_container.find_element_by_xpath('.//*[name()="svg"]/*[name()="svg"]')
    width = info_about_field.get_attribute("width")
    height = info_about_field.get_attribute("height")
    left_side = shotmap_container.find_element_by_xpath('.//*[name()="svg"]')
    shotmap_circles = left_side.find_elements_by_xpath('.//*[name()="circle"]')
    use_infos = left_side.find_elements_by_xpath('.//*[name()="use"]')

    driver.execute_script("arguments[0].scrollIntoView();", shotmap_container)

    # We need to remove top banner to be able to click
    try:
        driver.execute_script("""
        var l = document.evaluate('//*[@id="MatchFactsWrapper"]/div/div[1]/div[1]/section/div[2]/section[1]', document, null, XPathResult.FIRST_ORDERED_NODE_TYPE, null).singleNodeValue;
        l.parentNode.removeChild(l);
        """)
    except Exception as e:
        # print (e) # if would like to see full error
        print ("We do not have banner for that driver")
        pass

    all_shots = []
    for i in tqdm(range(len(use_infos)+len(shotmap_circles)-1)):
        shot_information = {}
        right_side = shotmap_container.find_element_by_xpath('.//div[contains(@class, "FullscreenShotInformationContainer")]')

        header_of_right_side = right_side.find_element_by_xpath('.//div[contains(@class, "commonBrowser")]')
        team_logo_url = right_side.find_elements_by_xpath('.//div[contains(@class, "PlayerIcon")]/img')[-1].get_attribute("src")
        player_url = header_of_right_side.find_element_by_xpath('.//a').get_attribute("href")
        player_name = header_of_right_side.find_element_by_xpath('.//a').text
        body_of_right_side = right_side.find_element_by_xpath('.//div[contains(@class, "ShotInfoContainer")]')

        minute = right_side.find_element_by_xpath('.//div[contains(@class, "BrowseContainer")]/div[contains(@class, "BrowseControllers")]/span').text[:-1]
        if " " in minute:
            minute_without_addition = float(minute.split(" ")[0])
        else:
            minute_without_addition = float(minute)
 
        left_of_body_of_right_side = body_of_right_side.find_element_by_xpath('.//div[contains(@class, "ShotInfo")]')
        right_of_body_of_right_side = body_of_right_side.find_element_by_xpath('.//div[contains(@class, "GoalGraphics")]')

        # ENHANCEMENT: That may be bit slow but for now, that is fine. We can revisit that.
        infos_from_left_of_body_of_right_side = {}
        for li in left_of_body_of_right_side.find_elements_by_xpath('.//ul/li'):
            spans =  li.find_elements_by_xpath('.//span')
            infos_from_left_of_body_of_right_side[spans[0].text] = spans[1].text

        xg_infos = right_of_body_of_right_side.find_element_by_xpath('.//div[contains(@class, "XGContainer")]').text
        xg_infos_dict = {xg_infos.split("\n")[1]:xg_infos.split("\n")[0],
                        xg_infos.split("\n")[3]:xg_infos.split("\n")[2]}

        cx = shotmap_circles[-1].get_attribute("cx")
        cy = shotmap_circles[-1].get_attribute("cy")

        shot_information["location"] = {"cx": float(cx), "cy": float(cy)}
        shot_information["type_and_result"] = infos_from_left_of_body_of_right_side
        shot_information["player_url"] = player_url
        shot_information["player_name"] = player_name
        shot_information["minute"] = eval(minute) # to convert 45 + 2' --> 47

        if meta_of_match["home_team_id"] in team_logo_url:
            shot_information["team_id_of_player"] = meta_of_match["home_team_id"]
            shot_information["team_side_of_player"] = "home"
        elif meta_of_match["away_team_id"] in team_logo_url:
            shot_information["team_id_of_player"] = meta_of_match["away_team_id"]
            shot_information["team_side_of_player"] = "away"
        else:
            raise Exception("id issue in shotmap")

        #shot_information["team_logo_url"] = team_logo_url # not necessary for now.
        shot_information["minute_without_addition"] = minute_without_addition
        for xg_key, xg_value in xg_infos_dict.items():
            # we have xG and XGOT here.
            if xg_value == "-":
                shot_information[xg_key] = None
            else:
                shot_information[xg_key] = float(xg_value)
        all_shots.append(shot_information)
        left_button, right_button = right_side.find_elements_by_xpath('.//div[contains(@class, "BrowseContainer")]/div[contains(@class, "BrowseControllers")]/button')[:]
        left_button.click()

    # In the following lines, we are checking that are we done correctly or not.
    # For that we are comparing 'minute' of the current shot in the page (which should be last one in the match)
    # and last appended shot's 'minute'. That last appended's time should be less than what we have in the page.
    minute = right_side.find_element_by_xpath('.//div[contains(@class, "BrowseContainer")]/div[contains(@class, "BrowseControllers")]/span').text[:-1]
    if " " in minute:
        minute_without_addition = float(minute.split(" ")[0])
    else:
        minute_without_addition = float(minute)
    if len(all_shots) > 0:
        if minute_without_addition > all_shots[-1]["minute_without_addition"]:
            print ("We are done!")
        else:
            print ("Possibly some shots are missing in the result. Please check that.")
    else:
        raise Exception ("There is no shots in the result")
    
    if sort == True:
        return all_shots[::-1]
    else:
        return all_shots
# example
import time
driver = None
url = "https://www.fotmob.com/match/3610281/matchfacts/manchester-city-vs-newcastle-united"
driver = open_the_url(url=url, driver=driver)
shotmap = get_shotmap(driver=driver)
```

**Comment by oseymour** [comment]
This is awesome! I'll finish my code and then compare. Have you profiled the time it takes to scrape each shot yet?

**Comment by hedonistrh** [comment]
For the code, which i shared, each shot takes around 0.7 seconds. 🐌 

I tried to use your code with the following code after copying [fotmob code](https://github.com/oseymour/ScraperFC/blob/main/code/ScraperFC/FotMob.py) into my iPython kernel.
```python
scraper = FotMob()
scraper.scrape_shots(match_link="https://www.fotmob.com/match/3610281/matchfacts/manchester-city-vs-newcastle-united") # as an example, use man-city vs new-united
```
With and without _headless_ option, however, both of them give following issue 
```bash
NoSuchElementException                    Traceback (most recent call last)
<ipython-input-7-ce63695ca956> in <module>
    162
    163 scraper = FotMob()
--> 164 scraper.scrape_shots(match_link="https://www.fotmob.com/match/3610281/matchfacts/manchester-city-vs-newcastle-united")

<ipython-input-7-ce63695ca956> in scrape_shots(self, match_link)
    107         while df.shape[0] < len(shots):
    108             new_row = pd.Series(dtype=object)
--> 109             new_row["minute"] = np.sum([int(i) for i in self.driver.find_element(
    110                 By.XPATH,
    111                 '//*[@id="MatchFactsWrapper"]/div/div[1]/div[4]/section/section/div[3]/div[1]/div/div[1]/div[2]/span'

/usr/local/lib/python3.9/site-packages/selenium/webdriver/remote/webdriver.py in find_element(self, by, value)
    974                 by = By.CSS_SELECTOR
    975                 value = '[name="%s"]' % value
--> 976         return self.execute(Command.FIND_ELEMENT, {
    977             'using': by,
    978             'value': value})['value']

/usr/local/lib/python3.9/site-packages/selenium/webdriver/remote/webdriver.py in execute(self, driver_command, params)
    319         response = self.command_executor.execute(driver_command, params)
    320         if response:
--> 321             self.error_handler.check_response(response)
    322             response['value'] = self._unwrap_value(
    323                 response.get('value', None))

/usr/local/lib/python3.9/site-packages/selenium/webdriver/remote/errorhandler.py in check_response(self, response)
    240                 alert_text = value['alert'].get('text')
    241             raise exception_class(message, screen, stacktrace, alert_text)
--> 242         raise exception_class(message, screen, stacktrace)
    243
    244     def _value_or_default(self, obj, key, default):

NoSuchElementException: Message: no such element: Unable to locate element: {"method":"xpath","selector":"//*[@id="MatchFactsWrapper"]/div/div[1]/div[4]/section/section/div[3]/div[1]/div/div[1]/div[2]/span"}
  (Session info: chrome=101.0.4951.64)
```

Currently I do not have enough time to debug but will try to check what is going on.

**Comment by oseymour** [comment]
0.7 is really good! My old code took around 50 seconds per shot. I've gotten mine down to around 0.3 seconds/shot, so I'll probably use mine when I release the FotMob module.

Again, the code on the repo is really old and was never tagged for release. It's not expected to work. I'll tag the FotMob release and announce it on twitter when it's done.

**Comment by hedonistrh** [comment]
Sounds great. 👌  I'll compare when you release it _(as profiling depends on the system so much)_ and if I find any speed-up, will share with you. 💯  Thanks!

**Comment by hedonistrh** [comment]
Hey Owen, I found the bug with my implementation. 😪   I was thinking that when you open the one of the match page, shotmap starts with the last shot of the game. _(This was the case before if i am not missing anything)_ But today, it is starting with the last shot which is 'goal'. So the code which I shared are able to still get all shots but sorting of that and logging message is not fully correct. I will probably share the fix tomorrow. 

AFAIK you are not using that but i will share just for future reference. ^-^
